### PR TITLE
Represent the absence of bias by `bias=false`

### LINKED ISSUE #1332: GPU error when using Zeros() as bias in Conv layer
When I try to use a Conv layer without bias (Julia: 1.5, CUDA: v1.3.3, Flux: v0.11.1), I have an error with GPU (all good with CPU).

Minimal example:

```julia
using Flux
using CUDA
using Flux: onehotbatch
using Flux: crossentropy
using Flux: convfilter, Zeros

CUDA.allowscalar(false)

batch = (Float32.(rand(32,32,3,4)), onehotbatch([8,5,9,2], 0:9)) |> gpu;
model = Chain(Conv(weight=convfilter((3,3), 3=>64), bias=Zeros()), 
	GlobalMaxPool(), flatten, Dense(64, 10), softmax) |> gpu;

loss(x, y) = crossentropy(model(x), y)
opt = ADAM()
ps = params(model);
gs = gradient(ps) do
	training_loss = loss(batch...)
	return training_loss
end
Flux.Optimise.update!(opt, ps, gs)
```

Error with julia debug level set to 2:

```julia
ERROR: a exception was thrown during kernel execution.
Stacktrace:
 [1] Bool at float.jl:73
 [2] convert at number.jl:7
 [3] setindex! at /home/afertin/.julia/packages/CUDA/dZvbp/src/device/array.jl:101
 [4] _setindex! at abstractarray.jl:1176
 [5] setindex! at abstractarray.jl:1153
 [6] broadcast_kernel at /home/afertin/.julia/packages/GPUArrays/eVYIC/src/host/broadcast.jl:62
ERROR: KernelException: exception thrown during kernel execution on device GeForce RTX 2080 Ti
Stacktrace:
 [1] check_exceptions() at /home/afertin/.julia/packages/CUDA/dZvbp/src/compiler/exceptions.jl:93
 [2] prepare_cuda_call() at /home/afertin/.julia/packages/CUDA/dZvbp/src/state.jl:85
 [3] context at /home/afertin/.julia/packages/CUDA/dZvbp/src/state.jl:142 [inlined]
 [4] cufunction(::GPUArrays.var"#broadcast_kernel#18", ::Type{Tuple{CUDA.CuKernelContext,CuDeviceArray{Bool,0,CUDA.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{},typeof(+),Tuple$
Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{0},Nothing,typeof(*),Tuple{Float64,Base.Broadcast.Extruded{CuDeviceArray{Bool,0,CUDA.AS.Global},Tuple{},Tuple{}}}},Base.Broadcast.Broadcast$d{CUDA.CuArrayStyle{0},Nothing,typeof(*),Tuple{Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{0},Nothing,typeof(-),Tuple{Int64,Float64}},Base.Broadcast.Broadcasted{CUDA.Cu$rrayStyle{0},Nothing,typeof(CUDA.culiteral_pow),Tuple{CUDA.CuRefValue{typeof(^)},Base.Broadcast.Extruded{CuDeviceArray{Float32,0,CUDA.AS.Global},Tuple{},Tuple{}},CUDA.CuRefValue{Val{2}$}}}}}},Int64}}; name::Nothing, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/afertin/.julia/packages/CUDA/dZvbp/src/compiler/execution.jl:293
 [5] cufunction(::GPUArrays.var"#broadcast_kernel#18", ::Type{Tuple{CUDA.CuKernelContext,CuDeviceArray{Bool,0,CUDA.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{},typeof(+),Tuple$Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{0},Nothing,typeof(*),Tuple{Float64,Base.Broadcast.Extruded{CuDeviceArray{Bool,0,CUDA.AS.Global},Tuple{},Tuple{}}}},Base.Broadcast.Broadcast$d{CUDA.CuArrayStyle{0},Nothing,typeof(*),Tuple{Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{0},Nothing,typeof(-),Tuple{Int64,Float64}},Base.Broadcast.Broadcasted{CUDA.Cu$rrayStyle{0},Nothing,typeof(CUDA.culiteral_pow),Tuple{CUDA.CuRefValue{typeof(^)},Base.Broadcast.Extruded{CuDeviceArray{Float32,0,CUDA.AS.Global},Tuple{},Tuple{}},CUDA.CuRefValue{Val{2}$}}}}}},Int64}}) at /home/afertin/.julia/packages/CUDA/dZvbp/src/compiler/execution.jl:293
 [6] #launch_heuristic#838 at /home/afertin/.julia/packages/CUDA/dZvbp/src/gpuarrays.jl:19 [inlined]
 [7] launch_heuristic at /home/afertin/.julia/packages/CUDA/dZvbp/src/gpuarrays.jl:17 [inlined]
 [8] copyto! at /home/afertin/.julia/packages/GPUArrays/eVYIC/src/host/broadcast.jl:66 [inlined]
 [9] copyto! at /home/afertin/.julia/packages/GPUArrays/eVYIC/src/host/broadcast.jl:76 [inlined]
 [10] materialize! at ./broadcast.jl:848 [inlined]
 [11] materialize!(::CuArray{Bool,0}, ::Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{0},Nothing,typeof(+),Tuple{Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{0},Nothing,typeof(*),Tuple{F
loat64,CuArray{Bool,0}}},Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{0},Nothing,typeof(*),Tuple{Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{0},Nothing,typeof(-),Tuple{I
nt64,Float64}},Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{0},Nothing,typeof(CUDA.culiteral_pow),Tuple{Base.RefValue{typeof(^)},CuArray{Float32,0},Base.RefValue{Val{2}}}}}}}}) at ./bro
adcast.jl:845
 [12] apply!(::ADAM, ::CuArray{Bool,0}, ::CuArray{Float32,0}) at /home/afertin/.julia/packages/Flux/05b38/src/optimise/optimisers.jl:176
 [13] update!(::ADAM, ::CuArray{Bool,0}, ::CuArray{Float32,0}) at /home/afertin/.julia/packages/Flux/05b38/src/optimise/train.jl:23
 [14] update!(::ADAM, ::Zygote.Params, ::Zygote.Grads) at /home/afertin/.julia/packages/Flux/05b38/src/optimise/train.jl:29
 [15] top-level scope at REPL[13]:1
```

### PR DESCRIPTION
**Pull request by mcabbott**
This continues the good work of #1379 to its logical conclusion, by making the storage match the API: `Dense(3,5; bias=false)` literally stores `randn(5,3), false, identity`. 

* This neatly drops out of `params`, is handled well by `loadparams!` & `destructure`, automatically.
* Since booleans are categorical in Zygote, they can be dropped completely in the gradient. ~~The code from https://github.com/FluxML/Zygote.jl/pull/838 is for now duplicated here for testing purposes.~~
* Maybe there's some tiny cost to doing `σ.(Wx .+ false)` during inference, but it's pretty hard to detect. 

Broadcasting `identity.(Wx)` is an easier cost to detect, since it copies the whole array. (Also only inference, removed by Zygote.) We could avoid both of these at the cost of a little complication, like some function `mapadd(σ, Wx, b)` to handle this, which I'm happy to write if there is interest. But nobody has thought this worth the complication so far --- which I read as saying that Flux values simplicity.

Smaller points:

* It has depwarns which catch both `Zeros()` and `Zeros(dims...)`.
* The function `create_bias` from #1379 is retained, and now also checks the size of the bias vector, if the user provides their own.
* `conv.jl` has fewer lines of code, by moving the `reshape(c.bias, map(_->1, c.stride)..., :, 1)` thing to a function, instead of repeating it.
* ~~There is some overlap with #1393, as this was written some time ago. This PR makes the type conversion noisy & explicit, although removing it entirely might be cleaner.~~
* I see no reason this should un-fix  #1332, but haven't tested on a GPU. Not sure if that's covered by CI.
* Closes #1392.

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -177,18 +177,20 @@ ones(dims...) = Base.ones(Float32, dims...)
 zeros(dims...) = Base.zeros(Float32, dims...)
 
 """
-    create_bias(shallcreate::Bool, iftrue, dims...) 
-    create_bias(x, ::Any...)
+    create_bias(bias::Bool, iftrue, dims...)
 
-Return a bias parameter for a layer.
+Return a bias parameter for a layer, based on keyword `bias=...`:
```
Comment (CarloLucibello) [inline src/utils.jl:182]: `bias` is an argument here, not a keyword

Comment (mcabbott) [reply src/utils.jl:182]: Indeed. It should say "the value passed to the constructor's keyword".

File: src/utils.jl [hunk 2]
```diff
@@ -177,18 +177,20 @@ ones(dims...) = Base.ones(Float32, dims...)
...
 
-Essentially handles the allowed input options for the `bias` keyword:
-    If `false`: Return the `Zeros` type which turns bias off.
-    If `true` : Return the result of `iftrue(dims)`.
-    If not a boolean, return self to handle the case of bias=somearray.
+* `bias=true` creates `iftrue(dims...)`, typically a zero array.
+* `bias=false` returns `false` to indicate no trainable bias,
+* `bias=randn(Float32,10)` uses the array provided.
```
Comment (CarloLucibello) [inline src/utils.jl:186]:
```suggestion
* if `bias isa AbtractArray` uses the array provided.
```

**Comment by CarloLucibello** [comment]
could you remove the conversion stuff? It's not related to this PR and I hope to merge #1393 soon

**Comment by mcabbott** [comment]
Yes, can do. Probably easiest as a rebase once that's merged, rather than trying to roll back smoothly enough that git doesn't notice. 

(It got added just because I landed in dispatch hell from the old implementation, somehow, and this was the quickest way out!)

**Comment by DhairyaLGandhi** [comment]
Since we've moved on from representing bias with `false` I would close this, feel free to open again if one thinks this needs a revisit

**Comment by mcabbott** [comment]
No, this is not done. Kindly try to keep things professional.

**Comment by CarloLucibello** [comment]
I still prefer this option to #1402 and to the solution we currently have. #1392 should be closed instead

**Comment by mcabbott** [comment]
No, this is not done. It is awaiting a rebase now that #1393 is in, as mentioned. Will someone who has sufficient permissions please re-open?

**Comment by CarloLucibello** [comment]
> Will someone who has sufficient permissions please re-open?

@DhairyaLGandhi please upgrade @mcabbott and @darsnack roles, don't force me to keep asking and asking

**Comment by DhairyaLGandhi** [comment]
Both of them are part of the same team as you or I.

**Comment by DhairyaLGandhi** [comment]
Either way, I think this approach hasn't been in line with the api changes that we have wanted to make here.

**Comment by CarloLucibello** [comment]
org members' permissions should be set to Write
https://docs.github.com/en/github/setting-up-and-managing-organizations-and-teams/setting-base-permissions-for-an-organization

**Comment by DhairyaLGandhi** [comment]
I am not inclined to deprecate Zeros or introduce a literal false, and it's somewhat surprising that folks here are not willing to consider the possibility of respecting existing code, their patterns, and the current solution also has questionable behavior with other tools. Added to that the solution in #1402 is generic enough to be used outside of this context without forcing any more burden on users as what this PR proposes and comes with more predictable behavior. 

Now I can understand that mcabbott wants to say that non existence of a bias means it doesn't exist, end of story. That's fine. But given a set of parameters how would you determine the model they came from? The skeleton model that we load the parameters into was typically expected to have this information prebaked, but once I tell you these params come from a typical resnet, with these slight tweaks to some biases being missing, without knowing them a priori, a singleton approach can't handle it, and would not be able to load the params correctly. Not to mention figuring out how to rewrite the model building code to make these distinctions. This is in part because this partial model specification was not embedded into the params. With a fake array approach, that is mitigated, and can be turned into trainable parameters at will, and used without any extra runtime penalties. The issues with param saving and GPU support were also fixed in #1402, so I don't understand why it's being pushed down this hard. Maybe it's more flexibility than we absolutely need at this moment, but it doesn't exactly come at extra cost either.

**Comment by mcabbott** [comment]
I think arguments in favour of Flux defining constant array types can be made in #1402. They would be strongest with references to deep learning models making use of them. 

> questionable behavior with other tools

Do you have a MWE?

> given a set of parameters how would you determine the model they came from?

You can't. A vector of numbers has less information than a model. That's why `destructure` needs to return a function which re-creates the model.

**Comment by darsnack** [comment]
Having been incidentally tagged on this PR, and now that I have spent over an hour catching up on the optional bias threads scattered all over, I have to say that I am _strongly_ in favor of the approach used in this PR over #1402. It is more maintainable and simpler (which as the original PR notes, is something Flux should continue to strive towards). I'll comment a bit more on #1402 as my thoughts are around `Zeros` not this PR.
